### PR TITLE
Add React helpers

### PR DESCRIPTION
**Pull request by nopelluhh**
## Changes
- Add `WTProvider`, `useTrack` and `useWT` hooks
- Reorganize code and tests to have clean split between `client` and `react` code

## Callouts
- `yarn.lock` was also deduped as part of adding new `react` dependencies.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -47,6 +47,92 @@ Events can be tracked without explicit parameters:
...
+
+`useWT` can be used if seeing the context event parameters is desired
+
+```tsx
+function List({ children }) {
+  const { track, params } = useWT();
+
+  doSomethingWithPageName(params.pageName);
+
+  return <button onClick={() => track("Tracked!")}>Track me!</button>;
```
Comment (bcharna) [inline README.md:115]: shouldn't an object be passed into `track`?

File: src/react.tsx [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    // eslint-disable-next-line @typescript-eslint/no-shadow
+    const { track, params } = useContext(WTContext);
+
+    return {
+      track,
+      params: params.current,
+    };
+  };
+
+  return { useWT: useWTContext, WTProvider, useTrack };
```
Comment (bcharna) [inline src/react.tsx:75]: don't have a strong opinion on this I suppose, but curious why we don't call the function `useWT` and avoid the rename here?

File: src/react.tsx [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+  params: React.RefObject<Partial<Params>>;
+  track: (overrides?: Partial<Params>) => void;
+};
+
+type Options<Params> = {
+  validateEvent?: (params: Partial<Params>) => boolean;
+};
+
+export const createProvider = <
+  Params extends Record<string | number, any> = WTEventParams
```
Comment (bcharna) [inline src/react.tsx:14]: do we want to allow number keys?

File: README.md [hunk 4]
```diff
@@ -47,6 +47,92 @@ Events can be tracked without explicit parameters:
...
+```
+
+### Hooks
+
+`useTrack` provides the most flexible interface to track an event in a context aware manner
+
+```tsx
+function List({ children }) {
+  // Parameters passed to `useTrack` will be merged into each call to `track`
+  const track = useTrack({ container: "list" });
```
Comment (bcharna) [inline README.md:93]: in the same vein as the above comment, it almost appears that you can use `useTrack` or `useWT` outside of a `WTProvider`. can we make the examples a bit more clear while still not having a ton of code to get the point across?

**Review by bcharna — COMMENTED** [review]
this is neat! just a handful of small comments/questions

File: README.md (hunk 4 above)
Comment (nopelluhh) [reply README.md:93]: I think code examples are best when they're very concise - I'm worried that adding extra context to each example will actually make them _less_ clear. How does amending the `React Integration` section to clarify how things must be composed sound?

File: README.md (hunk 1 above)
Comment (nopelluhh) [reply README.md:115]: 🤦 why yes, it should.

File: src/react.tsx (hunk 3 above)
Comment (nopelluhh) [reply src/react.tsx:14]: i don't think there's a particular reason to exclude them?

File: src/react.tsx (hunk 2 above)
Comment (nopelluhh) [reply src/react.tsx]: 06bfb29d391079e84b8c275fdce7a301bec08bc5

File: README.md [hunk 5]
```diff
@@ -47,6 +47,92 @@ Events can be tracked without explicit parameters:
...
+```
+
+### Context
+
+Context providers may be nested to build event parameters contextually:
+
+```tsx
+<WTProvider params={{ pageName: "home" }}>
+  <WTProvider params={{ container: "hero" }}>
+    {/* Events tracked here will receive the params { pageName: 'home', container: 'hero' } */}
```
[the remark this answers is not in the document]
Comment (nopelluhh) [reply README.md:84]: 45379b76cad258fdee92552dac7be0a09e647b03

File: README.md (hunk 4 above)
Comment (bcharna) [reply README.md:101]: that sounds good!

Comment (nopelluhh) [reply README.md:116]: updated here: 9d1f29a576bb26d934d99e1299abbcef8a86eaad

File: src/react.tsx (hunk 3 above)
Comment (bcharna) [reply src/react.tsx:14]: curious what a practical use case would be, but don't feel strongly about excluding them.

**Review by bcharna — APPROVED** [review]
great work on this!

**Pull request merged**